csrc/kernel_cache.cpp

```diff
@@ -1598,35 +1601,92 @@ std::optional<FusionKernelRuntime::HeuristicsPtr> FusionKernelRuntime::
+  FusionKernelRuntime::HeuristicsPtr heuristics =
+      std::make_unique<FusionHeuristics>(num_groups);
+
+  // Store metadata copy of arguments for ArgumentManager
+  KernelArgumentHolder args_metadata = copyMetadataArg(args);
+
+  // ArgumentManager manipulates the KernelArgumentHolder argument.
+  // We make another metadata copy for the PrecomputedValues.
+  KernelArgumentHolder complete_fusion_metadata_args(args_metadata);
+
+  ArgumentManager args_manager(
+      args_metadata, runtime_workspace_, segmented_fusion_->inputs());
+
+  // Follow group run order
+  for (int64_t group_id : c10::irange(num_groups)) {
+    auto group_to_run = runtime_workspace_.group_run_order.at(group_id);
+
+    // Create fusion for this segmented group
+    Fusion* fusion_to_run = group_to_run->getFusion();
+    NVF_ERROR(fusion_to_run != nullptr);
+    FusionGuard fg(fusion_to_run);
+
+    // Get input arguments for SchedulerRuntimeInfo
+    KernelArgumentHolder group_runtime_inputs;
+    for (auto input : group_to_run->inputs()) {
+      group_runtime_inputs.push(*args_manager.checkTensorMap(input));
     }
-    auto scheduler_entry = std::move(maybe_scheduler_entry.value());
-    if (!scheduler_entry->sameAs(
-            heuristics_->heuristicsList()[group_index].get())) {
-      return std::nullopt;
+
+    // Create PrecomputedValues initialized with original fusion inputs
+    auto evaluator_precomputed_values =
+        std::make_unique<PrecomputedValues>(fusion_to_run);
+    evaluator_precomputed_values->bindInputs(group_runtime_inputs);
+    evaluator_precomputed_values->bindValues(
+        group_to_run->getCompleteFusionInputs(), complete_fusion_metadata_args);
+    evaluator_precomputed_values->evaluate();
+
+    // Get all tensorviews for segmented fusion
+    std::vector<TensorView*> all_tvs_for_fusion_to_run =
+        ir_utils::allTvs(fusion_to_run);
+
+    SchedulerRuntimeInfo fusion_to_run_info(
+        fusion_to_run,
+        group_runtime_inputs,
+        evaluator_precomputed_values.get(),
+        all_tvs_for_fusion_to_run,
+        forced_index_type);
+
+    if (heuristics_ == nullptr) {
+      // Add new scheduler entry for this segmented group
+      heuristics->at(group_to_run->groupId()) =
+          segmented_fusion_->makeInitialSchedulerEntry(
+              group_to_run, fusion_to_run_info);
+    } else {
+      // Try to get scheduler entry
+      auto maybe_scheduler_entry =
+          group_to_run->getMaybeSchedulerEntry(fusion_to_run_info);
+      // If unavailable, then return std::nullopt
+      if (!maybe_scheduler_entry.has_value()) {
+        return std::nullopt;
+      }
+      // Check if this scheduler entry matches the previous entry for this
+      // segmented group. If no match, then return std::nullptr
+      auto scheduler_entry = std::move(maybe_scheduler_entry.value());
+      if (!scheduler_entry->sameAs(
+              heuristics_->at(group_to_run->groupId()).get())) {
+        return std::nullopt;
+      }
```

Comment:
nitpick: don't see much point in the move in the middle. maybe combine the two if checks together.